File: 0200-number-of-islands/0200-number-of-islands.py
from collections import deque
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:

        rows = len(grid)
        cols = len(grid[0])
        visited = set()
        directions =[(0, -1), (0, 1), (-1, 0), (1, 0)] #Up, Down, Left, Right
        islandCount = 0


        print(rows)
        print(cols)

        def bfs(row, col, visited, directions):
            queue = deque()
            queue.append((row, col, visited))

            while queue:
                row, col, visited = queue.pop()

                for nRow, nCol in directions:
                    newRow = row + nRow
                    newCol = col + nCol

                    if 0 <= newRow < rows and 0 <= newCol < cols and grid[newRow][newCol] == '1' and (newRow, newCol) not in visited:
                        visited.add((newRow, newCol))
                        queue.append((newRow, newCol, visited))

        for row in range(rows):
            for col in range(cols):
                if grid[row][col] == '1' and (row, col) not in visited:
                    print("Getting here")
                    islandCount += 1
                    visited.add((row, col))
                    bfs(row, col, visited, directions)

        return islandCount

        



        

File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) 
    {
        //Simple bfs
        int rows = grid.size();
        int cols = grid[0].size();

        std::vector<std::pair<int, int>> directions = {
            {-1, 0}, // up
            {0, -1}, // left
            {1, 0},  // down
            {0, 1}   // right
        };

        vector<vector<bool>> visited(rows, vector<bool>(cols, false));
        std::queue<std::pair<int, int>> queue;

        for (size_t i{}; i < rows; i++)
        {
            for (size_t j{}; j < cols; j++)
            {
                if (visited[i][j] || grid[i][j] == '0')
                {
                    //If found.
                    continue;
                }
                else
                {
                    std::cout << i << j << std::endl;
                    bfs(i, j, directions, visited, grid, queue, rows, cols);
                    res += 1;
                }
            }
        }

        return res;
    }

    void bfs(int row, int col, vector<std::pair<int, int>>& directions, vector<vector<bool>>& visited, vector<vector<char>>& grid, std::queue<std::pair<int, int>>& queue, int rows, int cols)
    {
        //Append
        queue.push({row, col});
        visited[row][col] = true;

        while (!queue.empty())
        {
            //Do something
            std::pair<int, int> curr = queue.front();
            queue.pop();

            //We have the coords
            int x = curr.first;
            int y = curr.second;

            for (auto& [newRow, newCol] : directions)
            {
                int nRow = newRow + x;
                int nCol = newCol + y;

                //Check if valid
                if (nRow >= 0 && nRow <= rows - 1 && nCol >= 0 && nCol <= cols - 1 && grid[nRow][nCol] == '1' && !visited[nRow][nCol] )
                {
                    queue.push({nRow, nCol});
                    visited[nRow][nCol] = true;
                }
            }
        }
    }

private:
    int res = 0;
};